mergify_engine/squash_pull.py

```diff
@@ -0,0 +1,119 @@
+# -*- encoding: utf-8 -*-
+#
+#  Copyright © 2021 Mergify SAS
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+import typing
+from mergify_engine import github_types
+from mergify_engine import context
+from mergify_engine.clients import http
+import dataclasses
+
+
+@dataclasses.dataclass
+class SquashFailure(Exception):
+    reason: str
+
+
+async def squash_pull(ctxt: context.Context) -> None:
+
+    if ctxt.pull["commits"] <= 1:
+        return
+
+    pull_number = ctxt.pull["number"]
+    head_branch = ctxt.pull["head"]["ref"]
+    head_owner = ctxt.pull["head"]["user"]["login"]
+    head_repo = ctxt.pull["head"]["repo"]["name"]
+    base_owner = ctxt.pull["base"]["user"]["login"]
+    base_repo = ctxt.pull["base"]["repo"]["name"]
+
+    try:
+        # get the branch
+        branch = typing.cast(
+            github_types.GitHubBranchRef,
+            await ctxt.client.item(
+                f"/repos/{head_owner}/{head_repo}/git/ref/heads/{head_branch}"
+            ),
+        )
+    except http.HTTPClientSideError:  # pragma: no cover
+        raise ValueError("ERROR: get the branch")
```

Comment:
Use your exception SquashFailure instead ValueError and forward more information:

The exception should provide what is failing ("branch not found") and why (`e.message`).

```python
except http.HTTPClientSideError as e:
     raise SquashFailure(f"branch not found: {e.message}")
```